packages/ai-jsx/src/batteries/docs.tsx

```diff
@@ -789,3 +806,56 @@ export async function DocsQA<ChunkMetadata extends Jsonifiable = Jsonifiable>(pr
     </ChatCompletion>
   );
 }
+
+const ResultSchema: z.ZodObject<any> = z
+  .object({
+    answer: z.string().describe("The answer to the user's question"),
+    sources: z.array(z.string()).describe('The title or URL of each document used to answer the question'),
+  })
+  .required({ answer: true });
+
+export type QAWithSourcesResult = z.infer<typeof ResultSchema>;
+
+/** A default component for formatting DocsQAWithSources results. */
+function DefaultQAResultFormatter(result: QAWithSourcesResult) {
+  if (result.sources?.length) {
+    return `${result.answer}\n\nSources:\n${result.sources.join('\n')}`;
+  }
+  return result.answer;
+}
+
+/**
+ * Similar to {@link DocsQA}, but encourages the LLM to return sources for its answer.
+ */
+export async function* DocsQAWithSources<ChunkMetadata extends Jsonifiable = Jsonifiable>(
+  props: DocsQAProps<ChunkMetadata>,
+  { render }: AI.ComponentContext
+) {
+  const chunks = await props.corpus.search(props.question, { limit: props.chunkLimit });
+  const chunkFormatter: (props: { doc: ScoredChunk<ChunkMetadata> }) => Node = props.chunkFormatter ?? DefaultFormatter;
+  const resultFormatter: (result: QAWithSourcesResult) => Node = props.resultFormatter ?? DefaultQAResultFormatter;
+
+  const stringifiedResult = (
+    <JsonChatCompletion schema={ResultSchema}>
+      <SystemMessage>
+        You are a trained question answerer. Answer questions truthfully, using only the document excerpts below. Do not
+        use any other knowledge you have about the world. If you don't know how to answer the question, just say "I
+        don't know." Here are the relevant document excerpts you have been given:
+        {chunks.map((chunk) => chunkFormatter({ doc: chunk }))}
+        And here is the question you must answer:
+      </SystemMessage>
+      <UserMessage>{props.question}</UserMessage>
+    </JsonChatCompletion>
+  );
+
+  const frames = render(stringifiedResult);
+  for await (const frame of frames) {
+    try {
+      yield resultFormatter(ResultSchema.parse(JSON.parse(frame)));
+    } catch (e) {
+      // The intermediate result isn't valid yet. Return it as is instead.
+      yield frame;
+    }
+  }
+  return resultFormatter(ResultSchema.parse(JSON.parse(await frames)));
+}
```

Comment:
I think this whole thing can be something like:

```tsx
return yield* render(stringifiedResult, {
  map: frame => {
    try { 
      return resultFormatter(ResultSchema.parse(JSON.parse(frame)));  
    } catch (e) { return frame }
  }
});
```

Reply:
Based on the docs (and a very quick read) for render, map only gets applied to intermediate results, so I don't think this quite works.

Reply:
Hmm @petersalas we should probably do something more ergonomic/obvious here.